runtime/gc_modron_startup/mmparse.cpp

```diff
@@ -559,18 +559,143 @@ xlpSubOptionsParser(J9JavaVM *vm, IDATA xlpIndex, XlpError *xlpError, UDATA *req
+			} else {
+				j9nls_printf(PORTLIB, J9NLS_INFO, J9NLS_GC_OPTIONS_LARGE_PAGE_SIZE_NOT_SUPPORTED_WITH_PAGETYPE, oldSize, oldQualifier, oldPageType, newSize, newQualifier, newPageType);
+			}
+		}
+	}
+}
+
+/**
+ * 
+ * Obtains and sets the preferred page size available on the system. 
+ * Returns true if the operation was successful.
+ * 
+ */
+bool
+gcSetPreferredPageSize(J9JavaVM *vm, MM_GCExtensions *extensions)
+{
+	UDATA defaultLargePageSize = 0;
+	UDATA defaultLargePageFlags = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	/* Get the default large page size of non pageable */
+	j9vmem_default_large_page_size_ex(0, &defaultLargePageSize, &defaultLargePageFlags);
+	
+	if (0 != defaultLargePageSize) {
+		extensions->requestedPageSize = defaultLargePageSize;
+		extensions->requestedPageFlags = defaultLargePageFlags;
+	} else {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * 
+ *  Sets the large page size used within the objectheap when the options -XX:+UseLargePages or -XX:LargePageSizeInBytes=<size> are used.
+ *  The value of the large page size will only be overwritten when these arguments are located in the right most index.
+ *  This is compared to the indices of -Xlp, -Xlp<size>, and -Xlp:objectheap:pagesize=<size>
+ * 
+ */
+static bool
+gcProcessXXLargePageObjectHeapArguments(J9JavaVM *vm)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(vm);
+	if (-1 == vm->largePageSizeRequested) {
+
+		UDATA defaultLargePageSize = 0;
+#if	defined(J9ZOS390)
+		UDATA defaultLargePageFlags = J9PORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE;
+#else /* J9ZOS390 */
+		UDATA defaultLargePageFlags = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
+#endif
+
+		j9vmem_default_large_page_size_ex(0, &defaultLargePageSize, &defaultLargePageFlags);
+
+		if (0 != defaultLargePageSize) {
+			extensions->requestedPageSize = defaultLargePageSize;
+			extensions->requestedPageFlags = defaultLargePageFlags;
+		} else {
+			j9nls_printf(PORTLIB, J9NLS_ERROR, J9NLS_GC_OPTIONS_SYSTEM_CONFIG_OPTION_NOT_SUPPORTED, "-XX:+UseLargePages");
+			return false;
+		}
+	} else {
+		UDATA largePageSizeRequested = vm->largePageSizeRequested;
+#if defined(J9ZOS390)
+		UDATA largePageFlagRequested = J9PORT_VMEM_PAGE_FLAG_PAGEABLE_PREFERABLE;
+#else /* J9ZOS390 */
+		UDATA largePageFlagRequested = J9PORT_VMEM_PAGE_FLAG_NOT_USED;
+#endif
+		verifyLargePageSize(vm, largePageSizeRequested, largePageFlagRequested, extensions);
+	}
+
+	return true;
+}
+
 /**
  * Find and consume -Xlp option(s) from the argument list.
  *
  * @param vm pointer to Java VM structure
  *
  * @return false if the option(s) are not consumed properly, true on success.
  */
-static bool
+static IDATA
 gcParseXlpOption(J9JavaVM *vm)
 {
 	MM_GCExtensions *extensions = MM_GCExtensions::getExtensions(vm);
-	bool rc = false;
+	IDATA rc = 0;
```

Comment:
Should it be `IDATA rc = -1;` ?

Reply:
In this method will return 0 if no `-Xlp`,` -Xlp<size>`, or `-Xlp:objectheap:pagesize=<size> `is used. 
The method will return -1 only if an error occurs. 

Looking at the code, it seems that 0 is a valid page index. Hence we need to reserve another negative integer to signify that no option was used. I can set rc = -2 to signify this case.

Reply:
You can see an implementation of this here: https://github.com/eclipse/openj9/pull/7476/commits/cad6beaec1cc182b1a5325f005ef62dba56d2810